feat(tusb_msc): Add device mount/unmount callback

roma-jam opened the pull request:
<!--
- Read and understand the project style guidelines (`CONTRIBUTION.md`).
- For Work In Progress Pull Requests, please use the Draft PR feature. See https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
- For a timely review/response, please avoid force-pushing additional commits if your PR has already received reviews or comments.
- Include screenshots for any CLI or UI changes.
- Keep PRs as small as possible; large PRs are difficult to review.
-->

## Description

<!--
- Please include a summary of the changes and the related issue.
- Also include the motivation (why this change) and context.
-->

<!-- 
- If you want to insert images (screenshots, diagrams, etc.), please format them:
    Bad link to the image (not formatted):   ![image](https://github.com/user-attachments/assets/ad3383cd-8f38-4d06-9ecf-3305e299307d)~~
    Good link to the image (formatted):       <img src="https://github.com/user-attachments/assets/ad3383cd-8f38-4d06-9ecf-3305e299307d" width=500px>
-->

## Related

<!--
- Use this format to link issue numbers: Fixes #123 - https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue-using-a-keyword
- Mention any other PRs related to this one.
- If there is related documentation, add the link here.
- If there is a public chat where changes in this PR were initiated, you can include the link here.
-->

## Testing

<!--
- Explain how you tested your change or new feature.
- If you tested changes in a clean environment (Docker container, new virtualenv, fresh Virtual Machine), state it here.
-->

---

## Checklist

Before submitting a Pull Request, please ensure the following:

- [ ] 🚨 This PR does not introduce breaking changes.
- [ ] All CI checks (GH Actions) pass.
- [ ] Documentation is updated as needed.
- [ ] Tests are updated or added as necessary.
- [ ] Code is well-commented, especially in complex areas.
- [ ] Git history is clean — commits are squashed to the minimum necessary.

roma-jam commented:
@tore-espressif @peter-marcisovsky @igi540 

Sorry for keep pushing after asking for the review. 

The new changes are for fixing the CI and taking into account the recent development guide for breaking change from esp-idf.

roma-jam commented:
As @tore-espressif noted, the merging of two init calls in one (for SDMMC and SPIFLASH) leads to the SDMMC code-related parts presence in the binary even if the SDMMC is not used, I reverted back the splitting and removes the enum. 

Now, the memory usage for `msc_storage` test application is next: 

|  Memory     | SPIFLAH only | SPIFLASH + SDMMC | SDMMC Delta |
| -------- | ------- | ---------- |  ----------- |
| Flash  | 270 728 | 334 074 | + 63 346 | 
| DIRAM | 90 920  | 91 560  |  + 640  |

... and no SDMMC-related files in the object files when `tinyusb_msc_storage_new_mmc()`* is not called from the code. 

*_I decided to name it with just mmc, as there is a possibility to use emmc as a MSC storage_

tore-espressif commented on device/esp_tinyusb/include/tinyusb_msc.h:
```diff
@@ -0,0 +1,219 @@
+ *    - ESP_ERR_INVALID_ARG: Invalid input argument
+ *    - ESP_ERR_NO_MEM: Not enough memory to initialize storage
+ */
+esp_err_t tinyusb_msc_storage_new_spiflash(const tinyusb_msc_config_t *config);
```
Would it be better to use the old API `tinyusb_msc_storage_init_spiflash(const tinyusb_msc_spiflash_config_t *config);` and just extend the configuration struct `tinyusb_msc_spiflash_config_t` ?

This would provide a lot more straight forward way for the users to upgrade to esp_tinyusb 2.0.0

If we can't just extend the struct we can deprecate some old struct members add new

roma-jam replied:
My main argument to have one config struct typedef `tinyusb_msc_config_t` instead of two for spiflash and sdmmc was having only one description with the not target specific context (sdmmc handle or wear levelling handle, or whatever else will be added later) instead of two, the same. 

So, I don't want to revert the changes and have two description back again, as it is a code duplication without any obvious advantages. 

Renaming the init function (`init` -> `new`) was made under the https://idf.gitlab-pages.espressif.cn:6688/chip-support-wiki/drivers/design-pattern-factory.html rules, where I applied the rule about creating the type inside (s_storage_handle), so that is why the new name has `*_new_*` in the naming. 

I was thinking about returning the handle (to fully comply the design pattern), because there is a technical possibility to create (I assume) two strorages on SPIFLASH and SDMMC and use them both, mapping to the different LUNs of MSC class. 
But I didn't come up with any valuable usecases for that, so I decided not to implement it. 

Although,  without usecases, I can add the handle (but unused) just as a API preparation for having several storages later, without introducing a breaking change. 

@tore-espressif 
I like the idea to introduce the storage handle now, but without the internal logic. 

Do you see any usecases for two storages (for example, one is mapped to SPIFLASH and one to SDMMC) working on two different LUNs on MSC Device class in parallel?

peter-marcisovsky commented on device/esp_tinyusb/tinyusb_msc.c:
```diff
@@ -0,0 +1,594 @@
+        break;
+    case FR_NO_FILESYSTEM:
+    case FR_INT_ERR:
+        ESP_LOGD(TAG, "Drive %s does not have a filesystem, need to format", drv);
```
Oh now I see.. sorry. Thx for clarification.

roma-jam commented on device/esp_tinyusb/tusb_msc_storage.c:
```diff
@@ -1,729 +0,0 @@
-esp_err_t tinyusb_msc_storage_init_spiflash(const tinyusb_msc_spiflash_config_t *config)
-{
-    assert(!s_storage_handle);
-    ESP_RETURN_ON_FALSE(CONFIG_TINYUSB_MSC_BUFSIZE >= CONFIG_WL_SECTOR_SIZE,
```
Just not intentionally. Sorry, thanks for pointing to that. 

I added the message, that project should be reconfigured in this case, because this problem is not possible to solve at run-time :(

roma-jam commented on device/esp_tinyusb/include/tinyusb_msc.h:
```diff
@@ -0,0 +1,219 @@
+ * @return
+ *   - ESP_OK: Callback set successfully
+ */
+esp_err_t tinyusb_msc_set_callback(tusb_msc_callback_t callback, void *arg);
```
Added the prototype and used it in the test_app.

roma-jam commented on device/esp_tinyusb/storage_sdmmc.c:
```diff
@@ -0,0 +1,108 @@
+        void *dest)
+{
+    sdmmc_card_t *card = (sdmmc_card_t *)ctx;
+    uint32_t sector_size = storage_sdmmc_get_sector_size(card);
```
Yep, just to avoid this impossible scenario, added the assert in `storage_sdmmc_get_sector_size()`.

roma-jam commented on device/esp_tinyusb/include/tinyusb_msc.h:
```diff
@@ -0,0 +1,286 @@
+
+#if (SOC_SDMMC_HOST_SUPPORTED)
+/**
+ * @brief Initialize TinyUSB MSC storage with SDMMC or EMMC
+ *
+ * This function initializes the TinyUSB MSC storage interface with MMC as the storage medium.
+ *
+ * @param[in] config Pointer to the configuration structure for TinyUSB MSC storage with SDMMC or EMMC
+ * @param[out] handle Pointer to the storage handle
+ *
+ * @return
+ *    - ESP_OK: Initialization successful
+ *    - ESP_ERR_INVALID_ARG: Invalid input argument
+ *    - ESP_ERR_NO_MEM: Not enough memory to initialize storage
+ */
+esp_err_t tinyusb_msc_storage_new_mmc(const tinyusb_msc_storage_config_t *config, tinyusb_msc_storage_handle_t *handle);
```
Correct spelling is SD/MMC.

So, uSD is a part of SD and eMMC is a part of MMC. 

So, previous naming was correct. 

```suggestion
 * @brief Initialize TinyUSB MSC SD/MMC storage
 *
 * This function initializes the TinyUSB MSC storage interface with SD/MMC as the storage medium.
 *
 * @param[in] config Pointer to the configuration structure for TinyUSB MSC storage with SD/MMC
 * @param[out] handle Pointer to the storage handle
 *
 * @return
 *    - ESP_OK: Initialization successful
 *    - ESP_ERR_INVALID_ARG: Invalid input argument
 *    - ESP_ERR_NO_MEM: Not enough memory to initialize storage
 */
esp_err_t tinyusb_msc_storage_new_sdmmc(const tinyusb_msc_storage_config_t *config, tinyusb_msc_storage_handle_t *handle);
```

roma-jam commented on device/esp_tinyusb/tinyusb_msc.c:
```diff
@@ -0,0 +1,806 @@
+        goto error;
+    }
+
+    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
+    size_t temp = 0;
+    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
+
+    if (__builtin_umul_overflow(lba, sector_size, &temp)) {
+        ESP_LOGE(TAG, "WRITE(10): Overflow lba");
+        goto error;
+    }
+
+    if (__builtin_uadd_overflow(temp, offset, &addr)) {
+        ESP_LOGE(TAG, "WRITE(10): Overflow addr");
+        goto error;
+    }
+
+    if (addr % sector_size != 0 || bufsize % sector_size != 0) {
+        ESP_LOGE(TAG, "WRITE(10): Invalid arguments");
+        goto error;
+    }
```
We need to verify overflow, only for SPI Flash storage.
So it is better to return the internal storage member storage_type and set it to TINYUSB_MSC_STORAGE_TYPE_SPIFLASH during the storage creation. 

Based on that, verify the overflow and report the error with status: 
```suggestion
    if (s_storage_handle->storage_type == TINYUSB_MSC_STORAGE_TYPE_SPIFLASH) {
        size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
        size_t temp = 0;
        size_t sector_size = p_msc_driver->dynamic.storage->sector_size;

        if (__builtin_umul_overflow(lba, sector_size, &temp)) {
            ESP_LOGE(TAG, "WRITE(10): Overflow address");
            goto error;
        }

        if (__builtin_uadd_overflow(temp, offset, &addr)) {
            ESP_LOGE(TAG, "WRITE(10): Overflow offset");
            goto error;
        }

        if (addr % sector_size != 0 || bufsize % sector_size != 0) {
            ESP_LOGE(TAG, "WRITE(10): Invalid arguments");
            goto error;
        }
    }
```

roma-jam commented:
Now, the memory usage for `msc_storage` test application is next: 

|  Memory     | SPIFLASH only | SPIFLASH + SDMMC | SDMMC Delta |
| -------- | ------- | ---------- |  ----------- |
| Flash  | 263 996 | 330 368 | + 66 372 | 
| DIRAM | 89 134  | 89 910 |  +  776 |

... and no SDMMC-related files in the object files when `tinyusb_msc_storage_new_sdmmc()` is not called from the code.

peter-marcisovsky commented on device/esp_tinyusb/test_apps/msc_storage/main/test_msc_storage.c:
```diff
@@ -0,0 +1,569 @@
+// ========================== TinyUSB General Device Descriptors ===============================
+//
+
+#define EPNUM_MSC       1
```
This define is unused. I presume it could be used in `TUD_MSC_DESCRIPTOR` below

roma-jam commented on device/esp_tinyusb/test_apps/msc_storage/main/test_msc_storage.c:
```diff
@@ -0,0 +1,569 @@
+    test_storage_wait_callback(TINYUSB_MSC_EVENT_FORMAT_REQUIRED);
+
+    // Format the drive
+    // tinyusb_msc_storage_format_drive();
```
I will add the logic, when the storage format feature will be implemented.

Check limitation, point 3.

roma-jam replied:
Will add the code after implementing the storage format feature.

roma-jam commented on device/esp_tinyusb/test_apps/msc_storage/main/test_msc_storage.c:
```diff
@@ -0,0 +1,569 @@
+
+    TEST_ASSERT_EQUAL(ESP_OK, tinyusb_driver_uninstall());
+    storage_deinit_spiflash(wl_handle);
+    tinyusb_msc_storage_delete(storage_hdl);
```
Added, thanks

roma-jam commented:
Hi @peter-marcisovsky,

I have addressed all the review notes from you, thanks a lot. 

I also updated the `tusb_msc_storage.h` file for simplifying the migration of the old code to the new version of esp_tinyusb. 

PTAL one more time! Thanks.

roma-jam commented:
@peter-marcisovsky 

Thanks for the review, feel free to approve, if you don't have anything else then.

> But I can see there are couple of more thing to be changed in the example.

Yes, there is one line is for tinyusb_msc class, so to solve it you need to remove the callbacks completely or update them to the new version. 

The rest - are the changes regarding the tinyusb driver itself. 

To proceed with the example, you need to: 
- solve the callbacks (by removing or updating them, commenting out the assigning to struct member will work)
- solve the tinyusb conflict: by using the default config macro `TINYUSB_DEFAULT_CONFIG()` 

I will provide this information in more detailed way in the README.

tore-espressif commented on device/esp_tinyusb/tinyusb_msc.c:
```diff
@@ -0,0 +1,871 @@
+
+    msc_storage_obj_t *s_storage_handle = p_msc_driver->dynamic.storage;
+
+    // As we defer the write operation to the TinyUSB task, we need to ensure that
+    // the address does not overflow for SPI Flash storage medium
+    // if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
+    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
+    size_t temp = 0;
+    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
+    ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
+    ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
+    // }
```
The comment says that we check overflows only for SPI flash, but it seems that we always check it.
Moreover, `addr` value is unused

Could you please check?

roma-jam replied:
Right, I commit the changes which I did to verify the Status with SD/MMC

tore-espressif commented:
> 5. [TODO]: Solve memory leakage of 100 bytes on one round in "erase -> format -> mount" -> "erase -> format -> mount" -> and so on.

It would be great if we could find where the leak is happening before merging this

roma-jam commented on device/esp_tinyusb/tinyusb_msc.c:
```diff
@@ -0,0 +1,871 @@
+
+    // As we defer the write operation to the TinyUSB task, we need to ensure that
+    // the address does not overflow for SPI Flash storage medium
+    // if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
+    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
+    size_t temp = 0;
+    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
+    ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
+    ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
+    // }
```
```suggestion
    if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
    size_t temp = 0;
    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
    ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
    ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
    }
```

roma-jam commented on device/esp_tinyusb/tinyusb_msc.c:
```diff
@@ -0,0 +1,871 @@
+
+    // As we defer the write operation to the TinyUSB task, we need to ensure that
+    // the address does not overflow for SPI Flash storage medium
+    if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
+    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
+    size_t temp = 0;
+    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
+    ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
+    ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
+    }
```
```suggestion
    if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
        size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
        size_t temp = 0;
        size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
        ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
        ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
    }
```
```suggestion
    if (s_storage_handle->medium->type == STORAGE_MEDIUM_TYPE_SPIFLASH) {
    size_t addr = 0; // Address of the data to be read, relative to the beginning of the partition.
    size_t temp = 0;
    size_t sector_size = p_msc_driver->dynamic.storage->sector_size;
    ESP_RETURN_ON_FALSE(!__builtin_umul_overflow(lba, sector_size, &temp), ESP_ERR_INVALID_SIZE, TAG, "overflow lba %lu sector_size %u", lba, sector_size);
    ESP_RETURN_ON_FALSE(!__builtin_uadd_overflow(temp, offset, &addr), ESP_ERR_INVALID_SIZE, TAG, "overflow addr %u offset %lu", temp, offset);
    }
```

roma-jam commented:
@tore-espressif 

> It would be great if we could find where the leak is happening before merging this

This PR is not a release yet, it is basically about the events API and refactoring of internal logic. 

There is another follow-up task: `[TODO]: Support Storage Format (only public API introduced)` when I will add the possibility to format the desired storage and provide the test for this. 

As this is happening during the format only, I can solve it during upcoming PR.
